cli/sql: print the execution time of statements

knz opened the pull request:
This patch adds the printing of query time after query results in the
CLI SQL shell. This helps users who use the CLI to try out queries
interactively and see which one performs faster.

Example output:
```
root@:26257/> create table test.kv(k int primary key, v int);
CREATE TABLE

Time: 20.596987ms

root@:26257/> create index vidx on test.kv(v);
CREATE INDEX

Time: 135.309993ms

root@:26257/> select * from test.kv;
+---+---+
| k | v |
+---+---+
+---+---+
(0 rows)

Time: 1.126903ms

root@:26257/>
```

This can be disabled/enabled with `\set show_time` / `\unset show_time`

The time display is automatically enabled when two conditions are met:
- the shell is interactive;
- the output format is `pretty`.

The first condition ensures that the results stay deterministic when
output is redirected, e.g. in tests. The second condition ensures that
a script that intends to have output formatted as valid HTML, CSV,
etc., does not get its output polluted by timing information.

Suggested by / informs #17353.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17358)
<!-- Reviewable:end -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but i think you'll need to replace time with timeutil.

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17358#-:-KqTp0_JOSRes2YMEHgv:bih2qz0)*
<!-- Sent from Reviewable.io -->

knz commented:
Good suggestion; done.

jordanlewis commented:
Can you put an output example? This seems great!

jordanlewis commented:
FWIW, this is how it works in `psql`:

```
jordan=# \timing
Timing is on.
jordan=# select 1;
 ?column?
----------
        1
(1 row)

Time: 0.298 ms
```

rjnn reviewed (approved):
LGTM, but seconding @jordanlewis I'd like to see what gets printed as well.

knz commented:
I extended the commit message with an example.

I chose for enabling this by default (instead of pg's client config setting) because I think this information is useful for discovering SQL in general and users should not have to make an extra step to get this benefit.

arjunravinarayan commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
I'm fine with making this a default.

---

Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17358#-:-KqTyUDTejS6xpTVCYrr:b-jwyt7t)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Personally I prefer to see a space between the number of rows output and the time spent, the way `psql` does it. But it's possible that's just because I'm used to seeing it that way.

knz commented:
I have added spaces before and after as suggested. Also made the printing configurable.

knz committed:
cli/sql: print the execution time of statements

This patch adds the printing of query time after query results in the
CLI SQL shell. This helps users who use the CLI to try out queries
interactively and see which one performs faster.

Example output:
```
root@:26257/> create table test.kv(k int primary key, v int);
CREATE TABLE

Time: 20.596987ms

root@:26257/> create index vidx on test.kv(v);
CREATE INDEX

Time: 135.309993ms

root@:26257/> select * from test.kv;
+---+---+
| k | v |
+---+---+
+---+---+
(0 rows)

Time: 1.126903ms

root@:26257/>
```

The time is only printed when the following two conditions are met:
- the shell is interactive;
- the output format is `pretty`.

The first condition ensures that the results stay deterministic when
output is redirected, e.g. in tests. The second condition ensures that
a script that intends to have output formatted as valid HTML, CSV,
etc., does not get its output polluted by timing information.

knz commented:
TFYRs!
